### PR TITLE
Add GPU implementation for tf.segment_mean

### PR DESCRIPTION
**Pull request by firejq**
According to @sanjoy 's [suggestion](https://github.com/tensorflow/tensorflow/pull/44436#pullrequestreview-523916485) in #44436 , I split out the implementation of segment mean and now proposal this pr.

Since these two pr have some common codes, and the other one is currently approved (but not yet merged), I have included all relevant codes in this pr. If the previous code is changed, I will make the corresponding modification here in time.

### CODE REVIEW

**Comment by firejq** [comment]
@sanjoy Could you please have a look?

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc [hunk 1]
```diff
@@ -126,6 +193,73 @@ __global__ void UnsortedSegmentCustomKernel(
   }
 }
 
+// SetSegmentFreq counts the frequency of occurrence of each element in
```
Comment (sanjoy) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:196]: I had a hard time following this function.  Have you considered phrasing `SetSegmentFreq` as a segment sum with input as an array of ones?  You could specialize the segment sum CUDA kernel to do this without having to create a temporary array.

If that idea does not work for some reason, please add a high level overview of the implementation.

Also see top level comment.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc [hunk 2]
```diff
@@ -126,6 +193,73 @@ __global__ void UnsortedSegmentCustomKernel(
...
+// 'segment_ids', and stores the result in each corresponding position
+// of 'segment_freq'.
+template <typename Index, int OuterDimTileSize>
```
Comment (sanjoy) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:199]: I'm not sure if `OuterDimTileSize` makes sense here, `segment_ids` is a vector right?

**Review by sanjoy — CHANGES_REQUESTED** [review]
Instead of writing something custom, have you considered implementing segment mean as:

```
SegmentMean(input, segment_ids) {
  sum = SegmentSum(input, segment_ids);
  counts = SegmentSum(tf.ones(input), segment_ids);
  result = ElementwiseDivide(sum, counts);
}
```

(You'd have to specialize the segment sum CUDA kernel to not need the intermediate allocation of `tf.ones`.)

**Comment by firejq** [comment]
@sanjoy Thanks for your review! Before making the corresponding changes, maybe we can further discuss the implementation of this op.


> Instead of writing something custom, have you considered implementing segment mean as:



Do you mean that we can implement `tf.segment_mean` just like the implementation of [tf.math.unsorted_segment_mean](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_ops.py#L4126), i.e. realize the equivalent function of `tf.segment_mean` through `tf.segment_sum` & `tf.ones` at the python level?

If so, I do have considered in fact. However, what defeated me was, that tensorflow had a stable cpu op implementation for `tf.segment_mean` already but this idea should remove the existing `tf.segment_mean` cpu implementation (right?). I think deleting an existing implementation may encounter greater verification costs, WDYT?

**Comment by sanjoy** [comment]
> Do you mean that we can implement `tf.segment_mean` just like the implementation of [tf.math.unsorted_segment_mean](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_ops.py#L4126), i.e. realize the equivalent function of `tf.segment_mean` through `tf.segment_sum` & `tf.ones` at the python level?

No, I meant doing it as an implementation detail of the `OpKernel`.

**Comment by gbaned** [comment]
@firejq  Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by firejq** [comment]
@sanjoy Sorry that I replied so late because some other jobs haunt me before.. 

> No, I meant doing it as an implementation detail of the `OpKernel`.

According to my previous test, although this implementation is simpler, the performance is not good due to the introduction of multiple kernels startup overhead, so I choose to implement it separately.  If we reach a consensus on this point, I would be nice to reopen and continue to improve this pr code. :)

**Comment by sanjoy** [comment]
> the performance is not good due to the introduction of multiple kernels startup overhead

I don't think you need to launch multiple kernels -- you can conceptually "fuse" the `tf.ones` into the `tf.segment_sum`, avoiding the temporary array and the kernel launch.

More concretely, I'm suggesting you parameterize the segment sum CUDA kernel on a `bool inputs_are_ones` template parameter which states whether the input is a real buffer or whether the kernel should pretend it is all ones (i.e. when `inputs_are_ones` is `true` the kernel won't load from the input (and the input buffer can just be null) and instead use `1`).  This can then be used to implement `SetSegmentFreq` more simply I think.

**Pull request closed without merging**